internal/gatewayapi/globalresources.go

```diff
@@ -18,25 +20,83 @@ import (
 const envoyTLSSecretName = "envoy"
 
 // ProcessGlobalResources processes global resources that are not tied to a specific listener or route
-func (t *Translator) ProcessGlobalResources(resources *resource.Resources, xdsIRs resource.XdsIRMap) error {
+func (t *Translator) ProcessGlobalResources(resources *resource.Resources, xdsIRs resource.XdsIRMap, gateways []*GatewayContext) error {
+	var errs error
+
 	// Get the envoy client TLS secret. It is used for envoy to establish a TLS connection with control plane components,
 	// including the rate limit server and the wasm HTTP server.
 	envoyTLSSecret := resources.GetSecret(t.ControllerNamespace, envoyTLSSecretName)
 	if envoyTLSSecret == nil {
-		return fmt.Errorf("envoy TLS secret %s/%s not found", t.ControllerNamespace, envoyTLSSecretName)
+		errs = errors.Join(errs, fmt.Errorf("envoy TLS secret %s/%s not found", t.ControllerNamespace, envoyTLSSecretName))
 	}
 
 	for _, xdsIR := range xdsIRs {
-		if containsGlobalRateLimit(xdsIR.HTTP) || containsWasm(xdsIR.HTTP) {
-			xdsIR.GlobalResources = &ir.GlobalResources{}
+		switch {
+		case t.MergeGateways:
+			svcClusterName := string(t.GatewayClassName)
+			svc := resources.GetServiceByLabels(map[string]string{OwningGatewayClassLabel: svcClusterName}, t.ControllerNamespace)
+			if svc == nil {
+				errs = errors.Join(errs, fmt.Errorf("envoy ServiceCluster for GatewayClass %s not found", t.GatewayClassName))
+				break
+			}
+
+			bRef := gwapiv1.BackendObjectReference{
+				Group:     GroupPtr(svc.GroupVersionKind().Group),
+				Kind:      KindPtr(svc.Kind),
+				Name:      gwapiv1.ObjectName(svc.Name),
+				Namespace: NamespacePtr(svc.Namespace),
+				Port:      PortNumPtr(svc.Spec.Ports[0].Port),
+			}
+			dst := t.processServiceDestinationSetting(svcClusterName, bRef, svc.Namespace, ir.AppProtocol(svc.Spec.Ports[0].Protocol), resources, resources.EnvoyProxyForGatewayClass)
+
+			xdsIR.GlobalResources = &ir.GlobalResources{
+				ProxyServiceClusters: []*ir.RouteDestination{{
+					Name:     dst.Name,
+					Settings: []*ir.DestinationSetting{dst},
+				}},
+			}
+		default:
+			for _, g := range gateways {
```

Comment:
this looks like O(n^2) where the top loop is IR and we are looping through gateways again, both of which have the same info

instead if you move it out into a separate section it will still be O(2n) but we can raise a follow up improve the logic

Reply:
Updated, please have another look